// nodejsapi/Constants.js
exports.Query = new function () {
    this.TABLE = "table";
    this.VIEW = "view";
    this.ASK = "ask";
    this.OSK = "osk";
    this._ID = "_id";
    this.COLUMNS = "columns";
    this.Columns = new function () {
        this.EXPRESSION = "expression";
        this.ALIAS = "alias";
        this.AGGERGATE = "aggregate";
        this.Aggregates = new function () {
            this.SUM = "sum";
            this.AVG = "avg";
            this.MAX = "max";
            this.MIN = "min";
            this.COUNT = "count";
        }
    };
    this.FILTER = "filter";
    this.Filter = new function () {
        this.Operator = new function () {
            this.QUERY = "$query";
            this.IN = "$in";
            this.WHEN = "$when";
            this.When = new function () {
                this.CONDITION = "$condition";
                this.FILTER = "$filter";
            }
        }
    }
    this.PARAMETERS = "parameters";
    this.ORDERS = "orders";
    this.Orders = new function () {
        this.ASC = "asc";
        this.DESC = "desc";
        this.ORDER = "$order";
        this.RECURSIVE = "$recursive"
        this.RESULT = "$result"
        this.Result = new function () {
            this.LEVEL = "level";
            this.NESTED = "nested";
        }
    }
    this.MAX_ROWS = "max_rows";
    this.CURSOR = "cursor";
    this.METADATA = "metadata";
    this.CHILDS = "childs";
    this.Childs = new function () {
        this.ALIAS = "alias";
        this.RELATED_COLUMN = "relatedcolumn";
        this.PARENT_COLUMN = "parentcolumn";
        this.QUERY = "query";
        this.USE_IN = "usein";
    }
    this.GROUPS = "groups";
    this.EXCLUDE_JOBS = "excludejobs";
    this.MODULE_PARAMETERS = "moduleparameters";
    this.KEEP_STRUCTURE = "keepstructure";
    this.Result = new function () {
        this.DATA = "data";
        this.METADATA = "metadata";
        this.CURSOR = "cursor";
        this.MODULE_RESULT = "moduleresult";
//        this.TOTAL_AGGREGATE_RESULT = "totalaggregateresult";
        this.ModuleResult = new function () {
            this.AGGREGATES = "aggregates";
            this.DATA = "data";
        }

        this.Tree = new function () {
            this.CHILDREN = "children";
            this.LEVEL = "_level";
            this.MODE = "_mode";
        }
    }
};


exports.Update = new function () {
    this.EXCLUDE_JOBS = "excludejobs";
    this.EXCLUDE_MODULES = "excludemodules";
    this.OPERATIONS = "operations";
    this.Operation = new function () {
        this.EXTRA_COLUMNS = "extracolumns";
        this.TYPE = "__type__";
        this.Type = new function () {
            this.INSERT = "insert";
            this.DELETE = "delete";
            this.UPDATE = "update";
            this.INSERT_IF_NOT_EXIST = "insertifnotexist";
        }
        this.OVERRIDE = "override";
    };

};
