// nodejsapi/lib/database/update.js
/**
 * Created with IntelliJ IDEA.
 * User: Administrator
 * Date: 7/25/13
 * Time: 4:02 PM
 * To change this template use File | Settings | File Templates.
 */

var HttpCommunication = require("../server/communication/HTTPCommunication.js")
var Constants = require("../../Constants.js");

var Update = function (table, options) {
    this.table = table;
    this.options = options;
    this.updates = [];
}

Update.prototype.insert = function (record) {
    var that = this;
    if (record instanceof Array) {
        record.forEach(function (rec) {
            that.updates.push(rec);
        })
    } else {
        that.updates.push(record);
    }
    return that;
}

Update.prototype.update = function (key, record) {
    record[Constants.Query._ID] = key;
    this.updates.push(record);
    return this;
}

Update.prototype.delete = function (key) {
    if (key instanceof Array) {
        var length = key.length;
        for (var i = 0; i < length; i++) {
            this.updates.push({_id:key[i], __type__:"delete"})
        }
    } else {
        this.updates.push({_id:key, __type__:"delete"})
    }
    return this;
}


Update.prototype.save = function (callback) {
    if (this.updates.length == 0) {
        callback(null, []);
        return;
    }
    var options = this.options || {};
    var update = {table:this.table, operations:this.updates};
    if (this.options && this.options.excludejobs) {
        update.excludejobs = this.options.excludejobs;
    }
    options[HttpCommunication.Services.Update.UPDATES] = JSON.stringify(update);
    HttpCommunication.call(HttpCommunication.Services.UPDATE, options, function (err, result) {
        this.updates = [];
        callback(err, result);
    });
}

module.exports = Update;
